### ISSUE TITLE
Harmony Hub disappears from Home App in TV Platform mode

### ISSUE THREAD

**Issue by DoctorExitus** [nicoduj/homebridge-harmony#82]
When I press on the Harmony Hub in the Home App or Control Center and try to change something, it disappears. Homebridge is still running fine. The Hub only reappears when I restart Homebridge. Closing and reopening the Home App does nothing. Also when I am in the Settings of the Hub in the Home App, I can see all my Activities and select which ones I want to show. But when I press on the Inputs button (bottom left) there are no Activities/Inputs. Mostly the Hub then disappears immediately afterwards. The Homebridge log file also says nothing new.

I am new to this, so if you need any more info I am glad if I can help. Thanks for your efforts btw. :)

This is my config:
"platforms": [
  {
    "platform": "HarmonyHubWebSocket",
    "name": "Harmony Hub",
    "hubIP": "192.168.XX.XX",
    "TVPlatformMode" : true,
    "mainActivity" : "Apple TV"
  }
]

**Comment by nicoduj** [comment]
Hi,

Could you tell me in which iOS beta you are ?

If I correctly understand the problem appears Chen you try to hide some input am I right ? If not everything is fine ? I didn’t check that use case so since there is a serialisation file that handle renaming of the inputs, maybe there is something wrong when you do this !

You can also send me your log in debug mode (-D) if you want further assistance , will check your use case however .





<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by DoctorExitus** [comment]
Hi,

I was on Public Beta 3, but now updated to PB4. The problem appears anytime I want to do something, not only when trying to hide inputs. 

I try to explain it more in detail: I start Homebridge. I go to the home app and add Homebridge as a new device. It shows the configuration screen where you can choose the room for Homebridge. I confirm. It then shows me the config screen for Harmony. All inputs are there and selected. I do not change anything and confirm. Both Homebridge and Harmony are in the default room now. I go there and it shows Harmony is active and on the "Apple TV" input, which is correct. I click on Harmony and go to inputs (lower left) and select another input. It changes correctly. I go back to the default room page and Harmony disappears. 
Then I restart Homebridge and go back to the home app and Harmony is back. But this time it only shows that it's turned on, without showing the input. In the settings view it shows all inputs, but when trying to change the input via the button on the lower left, it shows no inputs at all. Then it disappears again. 

I tried restarting homebridge and readding it to Home, but it always results in the same way. Sometimes it stays a little longer, sometimes I have to restart homebridge multiple times for it to come back. 


[debug log.txt](https://github.com/nicoduj/homebridge-harmony/files/2930739/debug.log.txt)

**Comment by nicoduj** [comment]
Ok, thank you for detailed information . Just in case your hub and homebridge have static ip ? I see you have a wol plugin what is it for ? Seems strange because the platform is static and there is no removal of the accessories once started .

Last questions : 

when you restart, does the room change ? 

Do you see the files in accessory folder of homebridge change when it happens ?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by DoctorExitus** [comment]
No they don't have a static IP. I use the homebridge iOS App. It installed Homebridge along with some plugins, but I have not configured them since I don't need them. WoL is one of them.

Yes, when I restart Homebridge, the Harmony device always changes to the same room as Homebridge.

**Comment by Midifreakz** [comment]
Hey @DoctorExitus 

I had the same problem the first time I tried TV-Mode. I fixed it when I deleted the Homebridge-Hub accessory from homekit and re-added homebridge to homekit all together.

**Comment by nicoduj** [comment]
Thanks for comment @Midifreakz

By the way static ip on the harmony hub is mandatory or you will have trouble when the up will change .

What’s strange is that I often delete / readd during dev process and never had such issue (only the missing inputs sometimes) .

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by DoctorExitus** [comment]
Hi @Midifreakz 

I tried deleting Homebridge from Homekit and even gave it a new identity, so that Homekit should see it as completely new. But it still happens. 

@nicoduj One thing I noticed is, that when I set up TVMode the first time in Homekit, the Harmony accessory was called "Harmony Hub". I changed it to "Harmony" because it was cut off on the accessory tile. However, when I delete everything now and readd Homebridge, the Harmony hub is already called "Harmony" without the "Hub". So maybe there is something that is persistent in the Homekit data that causes issues. In Homebridge the Harmony is still called "Harmony Hub". I only changed the name in Homekit.

**Comment by Midifreakz** [comment]
Are you using an apple tv or iPad as a homekit hub? perhaps try restarting that device too. helps me for 90% of homebridge/homekit problems.

**Comment by nicoduj** [comment]
@DoctorExitus indeed there is a file that stores names changes named harmonyPluginNames_xxxxxxxxxxxx at the root directory of your homebridge . It is why your name is back when you readd. Default name is the name in the config . But i doubt this is the cause of your issue 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by DoctorExitus** [comment]
I use an Apple TV 4K as Homekit hub. I just reset my whole Homekit by deleting the home and adding a new home and then adding all the devices back again. But the issue still persists.

I had a look into the error log (accessed via the Homebridge iOS App) and found some entries that refer to a "Visibility State". I don't know what that means but maybe it is of help to you, so I attached it.

One thing more thing I could try is to set up a new raspbian and install Homebridge via terminal, without the app. Maybe this Homebridge app does something strange that causes issues.

[error log.txt](https://github.com/nicoduj/homebridge-harmony/files/2931251/error.log.txt)

**Comment by nicoduj** [comment]
Don’t know this app ! Visibility state is always set to true from my side I think . Will check your log and this just in case this evening !

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by DoctorExitus** [comment]
Thanks for your effort, I sincerely appreciate it! 
I just went ahead and set up Homebridge on a fresh Raspbian installation. The problem is still the same. 

PS: I just thought, does the Apple TV need to be on tvOS 12.2 as well? It is the Homekit Hub and so far only my iPhone is on the Beta.

Version infos:
iOS 12.2 Public Beta 4
node.js v11.10.1
Homebridge v0.4.46
homebridge-harmony v0.4.7

And this is the config part for the plugin:
{
            "platform": "HarmonyHubWebSocket",
            "name": "Harmony",
            "hubIP": "192.168.178.24",
            "TVPlatformMode": true,
            "mainActivity": "Apple TV"
        }

**Comment by nicoduj** [comment]
Hi, my Apple TV is not, only my iPad is. On my iPhone the device is listed as unsupported but it is not an issue at all .

Do you have other plugins installed on your new setup ? 

As I said did you set a static ip to your hub ? It’s ip seems strange to me what kind of network do you use ?

**Comment by DoctorExitus** [comment]
Hi, I have a FritzBox router. They have the 192.168.178.XX subnet by default. So yes, all my devices are on the same subnet. I set both the Raspberry Pi and the Harmony Hub to static IP in the router settings. So they should always be given the same IP. I also installed homebridge-config-ui-x v3.9.6 because it makes it easier to restart or check Homebridge when I am not on the mac. Other than that no plugins.

**Comment by nicoduj** [comment]
Ok thanks, sorry for those questions I really don’t see what’s happening there ! I will check my node js version just in case , and go deeper in your log but really don’t know..

Oh yes one more question : when tv accessory disappear , do you still see the hub ? If yes in its properties does it list the tv device ? 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by DoctorExitus** [comment]
No problem, ask as many questions as you need :)

Yes the tv accessory disappears but the homebridge hub is still visible. In the properties, the tv device disappears as well.

**Comment by DoctorExitus** [comment]
If you mean the id in the file name, then it stays the same. I opened the file using nano, but it is just one  very long line, so I couldn't really see if something changed in it.

Yes when the device is there, it works as intended, just only a couple seconds before it disappears. And only after readding Homebridge in Homekit, it also shows the inputs. When I just restart Homebridge it shows no inputs and just says "Ein" (german for "turned on"). In the device properties however it shows the inputs.

**Comment by nicoduj** [comment]
So strange .... IS anyone by any chance have the same issu here ;) ? Don't really know where to look !

@DoctorExitus : I am interested into your file in persist folder (it is a json, you can format it as it with vscode for example but it is fine if you send it as it is).

**Comment by DoctorExitus** [comment]
I am knew to this, so how do I get the file that is on my Pi onto my Mac so I can upload it here? I only interact with the Pi via SSH and the terminal on my mac.

**Comment by nicoduj** [comment]
something like : 
scp your_username@remotehost.edu:foobar.txt /local/dir

replacing the files and path should do it

**Comment by DoctorExitus** [comment]
Alright, after a couple tries I got it. I hope it is useful. I also attached the AccessoryInfo file in case it is needed.

[AccessoryInfo.CC223DE3CE31.txt](https://github.com/nicoduj/homebridge-harmony/files/2932361/AccessoryInfo.CC223DE3CE31.txt)

[IdentifierCache.CC223DE3CE31.txt](https://github.com/nicoduj/homebridge-harmony/files/2932351/IdentifierCache.CC223DE3CE31.txt)

**Comment by DoctorExitus** [comment]
@nicoduj 

I think I found the issue. I still have an old iPad 3 which is on iOS 9. I disabled the iCloud access for Homekit on the iPad, then readded Homebridge to Homekit. It is now working and hasn't disappeared during the last 5 minutes. Maybe there is a conflict in how the different iOS versions handle Homekit Platforms and it disappeared after Homekit had synced with all Devices via iCloud.

**Comment by nicoduj** [comment]
wow that would be great news, but there was no chance I would have think about this ! Let me know, It could be very usefull for others !

**Comment by DoctorExitus** [comment]
Well, to be fair this wasn't really obvious at all. I just ran out of options, so I thought maybe it is iCloud related and then I thought of all devices that have access to Homekit via iCloud. Anyway, I have tested it for a while now and it seems that this indeed was the issue. So in case someone has the same problem, they likely have an active Apple device with an old iOS version.

Thank you very much for all your efforts and this amazing plugin. :)
